File: 백승우/2024-03-11/16928/16928.py
from sys import stdin
input = stdin.readline
from collections import deque

def bfs(n, m, ladder, snake):
  visited = set()
  q = deque([(0, 1)])

  while q:
    level, pos = q.popleft()
    visited.add(pos)

    if pos == 100:
      return level

    for dice_roll in range(6, 0, -1):
      next_pos = pos + dice_roll
      if next_pos > 100:
        continue

      if next_pos not in visited:
        if next_pos in ladder:
          next_pos = ladder[next_pos]
        elif next_pos in snake:
          next_pos = snake[next_pos]
        q.append((level + 1, next_pos))

  return -1


n, m = map(int, input().split())
ladder = {}
for _ in range(n):
  x, y = map(int, input().split())
  ladder[x] = y

snake = {}
for _ in range(m):
  u, v = map(int, input().split())
  snake[u] = v

min_moves = bfs(n, m, ladder, snake)

print(min_moves)

File: 백승우/2024-03-11/16928/16928.cpp
#include <iostream>
#include <queue>
#include <unordered_map>
#include <unordered_set>
using namespace std;

struct qInfo {
	int level;
	int pos;
};

int n, m;
unordered_map<int, int> ladder;
unordered_map<int, int> snake;
unordered_set<int> visited;

int bfs() {
	queue<qInfo> q;
	q.push({ 0, 1 });
	visited.insert(1);

	while (!q.empty()) {
		qInfo now = q.front();
		q.pop();

		if (now.pos == 100) return now.level;

		for (int i = 1; i <= 6; i++) {
			int nextPos = now.pos + i;

			if (nextPos > 100) continue;
			if (visited.find(nextPos) != visited.end()) continue;

			if (ladder.find(nextPos) != ladder.end()) {
				nextPos = ladder[nextPos];
			}
			else if (snake.find(nextPos) != snake.end()) {
				nextPos = snake[nextPos];
			}

			if (visited.find(nextPos) == visited.end()) {
				visited.insert(nextPos);
				q.push({ now.level + 1, nextPos });
			}
		}
	}

	return -1;
}

int main() {
	cin >> n >> m;
	for (int i = 0; i < n; i++) {
		int x, y;
		cin >> x >> y;
		ladder[x] = y;
	}
	for (int i = 0; i < m; i++) {
		int u, v;
		cin >> u >> v;
		snake[u] = v;
	}

	cout << bfs();

	return 0;
}
